### ISSUE TITLE
Is the intended behavior that firing a `close` event does not change `current`?

### ISSUE THREAD

**Issue by jgjgill** [toss/overlay-kit#44]
## Question

As I looked at the code, I had the following questions.

Neither `overlay.close()` nor `close` will change the `current`.
When I saw the name `current`, I thought it was related to the `overlay` that is currently `open`, but it's not.
Currently, `current` only changes in relation to `unmount`.

Is this the intended behavior?

## Test

I ran the following tests.

```tsx
import { act, render, screen } from '@testing-library/react';
import { useEffect, type PropsWithChildren } from 'react';
import { afterEach, describe, expect, it } from 'vitest';
import { OverlayProvider, useCurrentOverlay } from './context';
import { overlay } from './event';

afterEach(() => {
  overlay.unmountAll();
});

describe('overlay object', () => {
  it('overlay.close current test', async () => {
    const wrapper = ({ children }: PropsWithChildren) => <OverlayProvider>{children}</OverlayProvider>;

    const testContent1 = 'context-modal-test-content-1';
    const testContent2 = 'context-modal-test-content-2';
    const closeCurrent = 'close-current';

    let current: string | null = null;

    const Component = () => {
      current = useCurrentOverlay();

      useEffect(() => {
        overlay.open(
          ({ isOpen }) => {
            return <>{isOpen && <p>{testContent1}</p>}</>;
          },
          { overlayId: 'overlayId1' }
        );

        overlay.open(
          ({ isOpen, close }) => {
            return <>{isOpen && <p onClick={close}>{testContent2}</p>}</>;
          },
          { overlayId: 'overlayId2' }
        );
      }, []);

      return <div onClick={() => overlay.close(current!)}>{closeCurrent}</div>;
    };

    const renderComponent = render(<Component />, { wrapper });

    const testContentElement2 = await renderComponent.findByText(testContent2);
    const closeCurrenteElement = await renderComponent.findByText(closeCurrent);

    act(() => {
      testContentElement2.click(); // I expect `current` to change overlayId1.
      closeCurrenteElement.click(); // I expect `testContent1` to be closed.
    });

    expect(screen.queryByText(testContent1)).not.toBeInTheDocument();
    expect(screen.queryByText(testContent2)).not.toBeInTheDocument();
    expect(current).toBe(null);
  });
});
```

### Result

The `context-modal-test-content-1` still exists.

<img width="647" alt="스크린샷 2024-07-10 오후 7 17 57" src="https://github.com/toss/overlay-kit/assets/79239852/b49d897e-5b61-4449-a489-09a53281444f">

The `current` returns `overlayId2`, not `null`.

<img width="602" alt="스크린샷 2024-07-10 오후 7 18 35" src="https://github.com/toss/overlay-kit/assets/79239852/74ee276d-219d-45fd-904f-b9ba4524800f">

**Comment by jungpaeng** [comment]
I’ve given this matter a lot of thought.

The final decision to not have close affect current was based on the idea that close means the overlay is “hidden,” not fully closed.

When close is executed, the overlay changes isOpen to false, making it disappear from the screen, but it remains in memory and the DOM. Additionally, I assumed that executing close without unmount means the user is aware of this and intentionally keeps the overlay in memory. As a result, close does not modify current.

However, I’m open to the possibility that actual use cases might differ from my expectations. Could you share how you use current? If the fact that close does not change the current value is causing any issues, it would be helpful to know more about it.

**Comment by jgjgill** [comment]
Thank you for your quick feedback. 😆

> the overlay is “hidden,” not fully closed.

I understand. I felt the same way when I looked at the code. 🧐
but I think that it could be a little confusing for the user, so I left the issue open.

> Could you share how you use current?

Oh, I was just talking about what I was thinking while looking at the code, so I don't have an issue case yet 😇.
Still, I'll demo the issue I was thinking about while looking at the code tomorrow though!

Thank you for always being so detailed. 🙇‍♂️

**Comment by jgjgill** [comment]
Hi, @jungpaeng
I'll share what I thought was the issue situation. 🧐

[Demo](https://codesandbox.io/p/devbox/overlay-kit-demo-4vzvq6)

The conditions are as follows.

- Multiple modals exist on a single screen. (center, top, bottom, ...)
- The animation lasts for 2 seconds.
- You can `close` a modal via `current`. (Current Close Button)

Initially, I expected to close the modal every time I pressed the `Current Close Button`.
But in reality, the modal didn't close every time it was pressed.

https://github.com/toss/overlay-kit/assets/79239852/bd5f5ac2-f14f-4e29-9ab4-e3bcecc26683

But as you said, if `close` means `hidden`, then it stays until `current` disappears, so I understand that this is the intended behavior. 🙇‍♂️

## Sources of confusion

As I continued to think about it, I realized that my initial confusion came from using `close` and `current` together.

Please note that this is an arbitrary situation. 😇

If you close a modal that is not `current` and then close `current`, the two close together, as shown below.

https://github.com/toss/overlay-kit/assets/79239852/6d8f150b-8877-4051-997f-743bfcf6893b

`top open` -> `bottom open` (current) -> `top close`, `current close`

However, When trying to close modals using `current` only, only one will be closed.

So to summarize, I think the issue will arise when the user wants to close the modal whenever an event is fired using `close` and `current`.

**Comment by jungpaeng** [comment]
Thanks to the demo code and the video, I was able to understand it more easily.
This is my mistake. I didn’t consider that situation.

When thinking about the use cases for current, it would be used to display specific information on the screen depending on the currently open modal or to close that modal.
However, it is indeed awkward that the event does not trigger until the animation finishes.

It would be better to modify current as well when close is performed.

**Comment by jgjgill** [comment]
Oh, I'm glad I made a demo. 😆
When `current` also needed to be changed from `close`, I think of it like this.
It's okay to treat it as an idea!

- The `overlayOrderList` should not be changed for `REMOVE`.
- Duplicate 'customId' does not occur.

Based on the two conditions, I considered the index of `current` as follows.

- One less than the index corresponding to `customId` in `overlayOrderList`.

**reducer.ts**

```ts
    case 'CLOSE': {
      const closedCurrentIndex = state.overlayOrderList.findIndex((item) => item === action.overlayId);
      const currentIndex = closedCurrentIndex - 1;

      return {
        ...state,
        current: state.overlayOrderList[currentIndex] ?? null,
        overlayData: {
       ...
    }

    ...
    case 'REMOVE': {
      ...

      const copiedOverlayData = { ...state.overlayData };
      delete copiedOverlayData[action.overlayId];

      const updatedCurrent = state.current
        ? remainingOverlays.includes(state.current)
          ? state.current
          : remainingOverlays.at(-1) ?? null
        : null;

      return {
        current,
        overlayOrderList: remainingOverlays,
        overlayData: copiedOverlayData,
      };
    }
```

**Demo**

https://github.com/user-attachments/assets/7cb9ce13-220e-409d-9349-bc86c70c02d3

**Comment by jungpaeng** [comment]
I read through the code you provided carefully. I have a question about the code you wrote.

In `REMOVE`, when defining the `updatedCurrent` variable, you are checking if there is a current value and then removing it.

I assume this is to select the current overlay as C if overlays A, B, and C are open in that order and overlay B is unmounted.

However, even without this code, `remainingOverlays.at(-1)` would select C as the last item in the array.

Is there something I might be missing?

**Comment by jgjgill** [comment]
Oh, it works fine in that case as you said, so I tried to modify the code at first only in `CLOSE`.
But I found the following error case.
I'll leave you with [the branch](https://github.com/jgjgill/overlay-kit/tree/test/close-current) where I tested it. 🧐
Please note that I'm testing on a code base that I've arbitrarily modified. 😇

## Only change `CLOSE`

You can see that `current` gets temporarily weird when modals are closed sequentially, as shown below.
We expect `null` when all modals are removed, but after reaching `null` the first time, it changes to the disappearing `overlayId`s and then back to `null` again.

https://github.com/user-attachments/assets/e2f3da6a-5fa6-4b07-b049-61222a062956

This happens because `REMOVE` (`unmount`) changes `current` back to the last index of `remainingOverlays`.

## Error case

I think the error will occur in the following situations

- Create 4 Modals (or More)
- Close 3 (number doesn't matter)
- Firing a `close` or `unmount` event the moment `current` gets weird
- Nothing happened because we were trying to close a `current` that didn't already exist

https://github.com/user-attachments/assets/711b3731-6ea4-4448-970f-1f4654228663

## Conclusion

To solve this, I think `current` should not change when `current` exists inside `remainingOverlays`.

```tsx
const updatedCurrent = state.current
  ? remainingOverlays.includes(state.current)
    ? state.current // If `unmount` was executed after `close`
    : remainingOverlays.at(-1) ?? null // If you only run `unmount`, there is no `current` in `remainingOverlays`
  : null; // The case where `current` is `null`
```

https://github.com/user-attachments/assets/3b36081a-915e-46fc-aa1b-09d44852b0f1

https://github.com/user-attachments/assets/8ee94eb1-b1c7-4600-8836-7458ba2a3d32

If I've missed anything, feel free to let me know. 😆
Thanks. 🙇‍♂️

**Comment by jungpaeng** [comment]
Oh, thank you.

At first, I thought adjusting the overlayOrderList value at the `CLOSE` stage would solve the problem.
However, I realized that doing this would immediately remove the element from the DOM tree, leading to another bug.

Thanks to you, I think I can solve the problem cleanly.

**Comment by jungpaeng** [comment]
@jgjgill I have created PR #48.

I thought it would be good for you to review it before merging, so I’m tagging you for a quick check.

**Comment by jgjgill** [comment]
> this would immediately remove the element from the DOM tree, leading to another bug.

Yeah, I thought that It might be the issue [you were talking about](https://github.com/toss/overlay-kit/issues/37#issuecomment-2210844972). 🤔

I'm so glad I could help. I learned a lot too. 🙇‍♂️
